### PR TITLE
Feature/add clang tidy target

### PR DESCRIPTION
**Pull request by ClausKlein**
### Add check target to run clang-tidy while build
    
Add a _.clang-tidy_ config file too.
This may be used as prebuild target or like this:

```shell    
    cmake --build ../.build-doctest-Debug --target check |
     tee run-clang-tidy.log 2>&1
```
This may help to prevent warnings like this https://github.com/onqtam/doctest/issues/267

see too 
https://clang.llvm.org/extra/clang-tidy/Contributing.html#running-clang-tidy-on-llvm
and a **really good introduction**:
https://www.kdab.com/clang-tidy-part-1-modernize-source-code-using-c11c14/

### CODE REVIEW

File: scripts/cmake/common.cmake [hunk 1]
```diff
@@ -74,45 +74,47 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
...
-        add_compiler_flags(-pedantic)
-        add_compiler_flags(-pedantic-errors)
-        add_compiler_flags(-fvisibility=hidden)
-    endif()
+    # # The following options are not valid when clang-cl is used.
+    # if(NOT MSVC)
+    #     add_compiler_flags(-pedantic)
+    #     add_compiler_flags(-pedantic-errors)
+    #     add_compiler_flags(-fvisibility=hidden)
+    # endif()
```
Comment (onqtam) [inline scripts/cmake/common.cmake]: Sorry for the delay. Why is disabling these flags necessary?

File: scripts/cmake/common.cmake [hunk 2]
```diff
@@ -126,46 +128,53 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
...
+endif()
+set(CMAKE_CXX_EXTENSIONS ON)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compiler_flags(-Wall)
+    add_compiler_flags(-Wextra)
+    add_compiler_flags(-Wpedantic)
+
     add_compiler_flags(-Weverything)
```
Comment (onqtam) [inline scripts/cmake/common.cmake]: ```-Weverything``` should include whatever is in ```-Wall``` and ```-Wextra```

File: scripts/cmake/common.cmake [hunk 3]
```diff
@@ -126,46 +128,53 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
...
+if(NOT DEFINED CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_STANDARD 11)
+endif()
+set(CMAKE_CXX_EXTENSIONS ON)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compiler_flags(-Wall)
+    add_compiler_flags(-Wextra)
+    add_compiler_flags(-Wpedantic)
```
Comment (onqtam) [inline scripts/cmake/common.cmake]: ```-Weverything``` should include whatever is in ```-Wall``` and ```-Wextra```

File: scripts/cmake/common.cmake [hunk 4]
```diff
@@ -74,45 +74,47 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
...
-    add_compiler_flags(-Wdisabled-optimization)
-    add_compiler_flags(-Weffc++)
-    add_compiler_flags(-Winvalid-pch)
-    add_compiler_flags(-Wmissing-declarations)
-    add_compiler_flags(-Woverloaded-virtual)
-    add_compiler_flags(-Wunused-but-set-variable)
-    add_compiler_flags(-Wunused-result)
+    add_compiler_flags(-Wpedantic)
+
+    # add_compiler_flags(-fdiagnostics-show-option)
```
Comment (onqtam) [inline scripts/cmake/common.cmake]: All these flags aren't covered by ```-Wall``` and ```-Wextra``` - GCC doesn't have anything like the ```-Weverything``` which clang has...

File: scripts/run-clang-tidy.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
+#!/usr/bin/env python
```
Comment (onqtam) [inline scripts/run-clang-tidy.py]: I guess this file is copied from the LLVM project (or somewhere else) - does it need to be in the doctest repository? Can't it be used from wherever clang-tidy is located? what about the license?

**Comment by onqtam** [comment]
I haven't maintained the static analysis build on Travis for quite a while... I haven't got the time for that.

Perhaps I should remove OCLint and maybe even cppcheck - leaving just clang-tidy? not sure...

Is this PR meant to be just the first phase (fixing the infrastructure), followed by actually cleaning/suppressing the clang-tidy problems?

File: scripts/cmake/common.cmake (hunk 3 above)
Comment (ClausKlein) [reply scripts/cmake/common.cmake]: yes, bad what is about apple-clang?

File: scripts/run-clang-tidy.py (hunk 5 above)
Comment (ClausKlein) [reply scripts/run-clang-tidy.py]: Not shure, but not every distribution install it (windows, mac ports, ...)
Debian does.

File: scripts/cmake/common.cmake (hunk 1 above)
Comment (ClausKlein) [reply scripts/cmake/common.cmake]: Some of the gcc flags are not known to clang-tidy and it complain it
The compile_comands.json generated from cake file is parced by clang-tidy.

My hope is -Wpedantic activate them all

**Comment by cjdb** [comment]
What is the status of this work? I am pushing for a university class to adopt both clang-tidy and Doctest in their labs, so it'd be great to know if this is possible before the class starts.

**Comment by onqtam** [comment]
@cjdb Currently I don't have the capacity to review this - I'm mostly focusing on fixing critical bugs and not producing any warnings with the supported compilers, but static analysis has fallen behind in priorities. You may test what's in this PR - it might be good enough.

**Comment by onqtam** [comment]
reopening - accidentally deleted the dev remote branch and github auto-closed this PR

**Comment by onqtam** [comment]
I'm closing this - renewed clang-tidy efforts ought to go in new PRs

**Pull request closed without merging**